daemon/containerd/image_exporter.go

```diff
@@ -322,81 +314,3 @@ func (i *ImageService) LoadImage(ctx context.Context, inTar io.ReadCloser, outSt
 
 	return nil
 }
-
-// getBestDescriptorForExport returns a descriptor which only references content available locally.
-// The returned descriptor can be:
-// - The same index descriptor - if all content is available
-// - Platform specific manifest - if only one manifest from the whole index is available
-// - Reduced index descriptor - if not all, but more than one manifest is available
-//
-// The reduced index descriptor is stored in the content store and may be garbage collected.
-// It's advised to pass a context with a lease that's long enough to cover usage of the blob.
-func (i *ImageService) getBestDescriptorForExport(ctx context.Context, indexDesc ocispec.Descriptor) (ocispec.Descriptor, error) {
-	none := ocispec.Descriptor{}
-
-	if !containerdimages.IsIndexType(indexDesc.MediaType) {
-		err := fmt.Errorf("index/manifest-list descriptor expected, got: %s", indexDesc.MediaType)
-		return none, errdefs.InvalidParameter(err)
-	}
-	store := i.client.ContentStore()
-	children, err := containerdimages.Children(ctx, store, indexDesc)
-	if err != nil {
-		if cerrdefs.IsNotFound(err) {
-			return none, errdefs.NotFound(err)
-		}
-		return none, errdefs.System(err)
-	}
-
-	// Check which platform manifests have all their blobs available.
-	hasMissingManifests := false
-	var presentManifests []ocispec.Descriptor
-	for _, mfst := range children {
-		if containerdimages.IsManifestType(mfst.MediaType) {
-			available, _, _, missing, err := containerdimages.Check(ctx, store, mfst, nil)
-			if err != nil {
-				hasMissingManifests = true
-				log.G(ctx).WithField("manifest", mfst.Digest).Warn("failed to check manifest's blob availability, won't export")
-				continue
-			}
-
-			if available && len(missing) == 0 {
-				presentManifests = append(presentManifests, mfst)
-				log.G(ctx).WithField("manifest", mfst.Digest).Debug("manifest content present, will export")
-			} else {
-				hasMissingManifests = true
-				log.G(ctx).WithFields(log.Fields{
-					"manifest": mfst.Digest,
-					"missing":  missing,
-				}).Debug("manifest is missing, won't export")
-			}
-		}
-	}
-
-	if !hasMissingManifests || len(children) == 0 {
-		// If we have the full image, or it has no manifests, just export the original index.
-		return indexDesc, nil
-	} else if len(presentManifests) == 1 {
-		// If only one platform is present, export that one manifest.
-		return presentManifests[0], nil
-	} else if len(presentManifests) == 0 {
-		// Return error when none of the image's manifest is present.
-		return none, errdefs.NotFound(fmt.Errorf("none of the manifests is fully present in the content store"))
```

Comment:
Wondering if we want to have a utility / check around to allow returning a warning.

I think it makes sense to allow exporting "whatever is in the store" (even if partial content), but possibly we want to have;

- a warning
- an option to create a new variant of the image with only the available variants (and thus; new digest)

Not exactly sure yet what the "UX" for it would be, and if that would require adding an option to the API ("export policy" or whatever)